crates/esthri/src/tempfile.rs

```diff
@@ -52,6 +52,10 @@ impl TempFile {
         self.file.as_mut().unwrap()
     }
 
+    pub fn get_file(&mut self) -> File {
```

Comment:
I'm pretty sure this is wrong, but I'm not sure how to make it work with the `'static` lifetime needed for the upload. I think getting the File like this could run into issues as I think the temp file it's pointed to could get deleted from under it? Need to look into it further, just did the minimum to get it compiling but would welcome any ideas.

Reply:
Hmm I believe you're pretty much spot on here. I think what you have will work fine, fwiw. In order to call `take_std_file` or `take_file` (I just renamed `get_file` -> `take_file` to be consistent with the other method that takes the file), you just have to make sure that the rest of the `TempFile` you took the file from stays alive as long as you are using the file, so for example where we use it - 

```rust
let (mut tmp, size) = compress_to_tempfile(path).await?;
upload_from_reader(
    s3,
    bucket,
    key,
    tmp.get_file(),
    size,
    Some(crate::compression::compressed_file_metadata()),
)
.await?;
Ok(())
```

`tmp` should be alive for the whole duration of `upload_from_reader`, and after that we don't need it so were good.

I agree that it's not super clean, but I don't think there is a way around it due to the static constraint you mention. In order to support a non static reader I think we'd need `ByteStream` to look like (abbreviated slightly) 

```
pub struct ByteStream<'a> {
    inner: Pin<Box<dyn Stream<Item = Result<Bytes, io::Error>> + Send + 'a>>,
}
```

As to why it's not set up that way I'm not sure. It might be a limitation of the underlying http library, hyper, or it might have just been a design choice of the rusoto authors. I'm curious if the new aws sdk also has this constraint

Reply:
Makes sense! I guess the only downside is something like this would still compile but would cause issues?

```
let tmp: TemporaryFile; // temp file that already has things written to the file on disk
let file_handle = tmp.take_file().unwrap()
drop(tmp) // dropping causes the temp file to be deleted on disk
// file_handle is still valid even though the temporary file it points to has been dropped & deleted
```

..which I suppose is probably why the compiler doesn't like using references here.

It is kinda annoying that the `static` stuff means it's hard to get the nice advantages we'd otherwise get by using references pointed to a `TempPath`. :disappointed: